### PR TITLE
refactor `flutter upgrade` to be 2 part, with the second part re-entrant

### LINKED ISSUE #33890: VersionCheckError during `flutter upgrade` on macOS 10.15 beta (Catalina)
Flutter is not currently supported on macOS 10.15, but adding this issue for awareness.

Flutter crash report; please file at https://github.com/flutter/flutter/issues.

## command

flutter upgrade

## exception

VersionCheckError: VersionCheckError: Command exited with code -9: git log -n 1 --pretty=format:%ad --date=iso
Standard error: 

```
#0      _runSync (package:flutter_tools/src/version.dart:521:5)
#1      FlutterVersion._latestGitCommitDate (package:flutter_tools/src/version.dart:141:12)
#2      FlutterVersion.frameworkCommitDate (package:flutter_tools/src/version.dart:132:37)
#3      FlutterVersion.toString (package:flutter_tools/src/version.dart:109:99)
#4      UpgradeCommandRunner.updatePackages (package:flutter_tools/src/commands/upgrade.dart:190:32)
<asynchronous suspension>
#5      UpgradeCommandRunner.runCommand (package:flutter_tools/src/commands/upgrade.dart:90:11)
<asynchronous suspension>
#6      UpgradeCommand.runCommand (package:flutter_tools/src/commands/upgrade.dart:47:32)
<asynchronous suspension>
#7      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:488:18)
#8      _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:77:64)
#9      _rootRunUnary (dart:async/zone.dart:1132:38)
#10     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#11     _FutureListener.handleValue (dart:async/future_impl.dart:126:18)
#12     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:639:45)
#13     Future._propagateToListeners (dart:async/future_impl.dart:668:32)
#14     Future._complete (dart:async/future_impl.dart:473:7)
#15     _SyncCompleter.complete (dart:async/future_impl.dart:51:12)
#16     _AsyncAwaitCompleter.complete.<anonymous closure> (dart:async-patch/async_patch.dart:33:20)
#17     _rootRun (dart:async/zone.dart:1124:13)
#18     _CustomZone.run (dart:async/zone.dart:1021:19)
#19     _CustomZone.bindCallback.<anonymous closure> (dart:async/zone.dart:947:23)
#20     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#21     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#22     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:115:13)
#23     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:5)
```

## flutter doctor

```
encountered exception: VersionCheckError: Command exited with code -9: git log -n 1 --pretty=format:%ad --date=iso
Standard error: 

#0      _runSync (package:flutter_tools/src/version.dart:521:5)
#1      FlutterVersion._latestGitCommitDate (package:flutter_tools/src/version.dart:141:12)
#2      FlutterVersion.frameworkCommitDate (package:flutter_tools/src/version.dart:132:37)
#3      FlutterVersion.frameworkDate (package:flutter_tools/src/version.dart:94:31)
#4      FlutterValidator.validate (package:flutter_tools/src/doctor.dart:468:127)
<asynchronous suspension>
#5      Doctor.startValidatorTasks (package:flutter_tools/src/doctor.dart:134:52)
#6      Doctor.diagnose (package:flutter_tools/src/doctor.dart:205:41)
<asynchronous suspension>
#7      _doctorText.<anonymous closure> (package:flutter_tools/runner.dart:188:26)
#8      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:29)
<asynchronous suspension>
#9      _rootRun (dart:async/zone.dart:1124:13)
#10     _CustomZone.run (dart:async/zone.dart:1021:19)
#11     _runZoned (dart:async/zone.dart:1516:10)
#12     runZoned (dart:async/zone.dart:1463:12)
#13     AppContext.run (package:flutter_tools/src/base/context.dart:152:18)
<asynchronous suspension>
#14     _doctorText (package:flutter_tools/runner.dart:187:19)
<asynchronous suspension>
#15     _createLocalCrashReport (package:flutter_tools/runner.dart:165:32)
<asynchronous suspension>
#16     _handleToolError (package:flutter_tools/runner.dart:120:33)
<asynchronous suspension>
#17     run.<anonymous closure> (package:flutter_tools/runner.dart:64:20)
<asynchronous suspension>
#18     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:153:29)
<asynchronous suspension>
#19     _rootRun (dart:async/zone.dart:1124:13)
#20     _CustomZone.run (dart:async/zone.dart:1021:19)
#21     _runZoned (dart:async/zone.dart:1516:10)
#22     runZoned (dart:async/zone.dart:1463:12)
#23     AppContext.run (package:flutter_tools/src/base/context.dart:152:18)
<asynchronous suspension>
#24     runInContext (package:flutter_tools/src/context_runner.dart:54:24)
<asynchronous suspension>
#25     run (package:flutter_tools/runner.dart:51:10)
#26     main (package:flutter_tools/executable.dart:62:9)
<asynchronous suspension>
#27     main (file:///Users/timsneath/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)
#28     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:298:32)
#29     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:171:12)
```

The log seems to complain about this line: `git log -n 1 --pretty=format:%ad --date=iso`, but the command here works from my terminal.

### PR DESCRIPTION
**Pull request by christopherfujino**
Currently, if you run flutter upgrade on macOS Catalina and the upgrade path has a Dart SDK revision change, you will get an exception with a misleading VersionCheckError: Command exited with code -9: git log -n 1 --pretty=format:%ad --date=iso. The actual cause is calling the Flutter tool re-entrantly, which deletes the current Dart SDK (which the flutter upgrade process is using), after which any forked sub-processes would be killed (code -9 from the VersionCheckError message) by the macOS anti-malware software. See #33890 and https://github.com/dart-lang/sdk/issues/37662 for more info.

This PR breaks up the `flutter upgrade` flow into two parts, the second half invoked re-entrantly via the hidden flag `flutter upgrade --continue`. This fixes the issue because after there are no steps following the re-entrant call of `flutter upgrade --continue` from the first dart process.

## Description

*Replace this paragraph with a description of what this PR is doing. If you're modifying existing behavior, describe the existing behavior, how this PR is changing it, and what motivated the change. If you're changing visual properties, consider including before/after screenshots (and runnable code snippets to reproduce them).*

## Related Issues

Fixes https://github.com/flutter/flutter/issues/33890

## Tests

I re-wrote the existing `upgrade_test.dart` file to accommodate the new re-entrant command.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Can you add some more context to the PR description on how this helps to resolve the underlying issue for Catalina?

**Comment by christopherfujino** [comment]
> Can you add some more context to the PR description on how this helps to resolve the underlying issue for Catalina?

Done.

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/commands/upgrade.dart [hunk 1]
```diff
@@ -20,12 +20,19 @@ import 'channel.dart';
...
+    argParser
+      ..addFlag(
+        'force',
+        abbr: 'f',
+        help: 'Force upgrade the flutter branch, potentially discarding local changes.',
+        negatable: false,
+      )
+      ..addFlag(
+        'continue',
+        hide: true,
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/upgrade.dart]: we should expose this in verbose mode.

**Pull request merged**